### PR TITLE
Add the rest of schema swap code.

### PR DESCRIPTION
**Pull request by pivanof**
This code is still not tested besides simple compilation, but it is representing
now the full schema swap process from the beginning to the end. The code also
is not well-versed on handling different errors, so it mostly represents only
the happy path of the process.

Next up after this will be wiring the code into a way to call it and actually
testing it.

### CODE REVIEW

File: go/vt/schemamanager/schemaswap/schema_swap.go [hunk 1]
```diff
@@ -75,6 +90,290 @@ type shardSchemaSwap struct {
...
+	currentSQL string
+}
+
+// Run is the main entry point of the schema swap process. It drives the process from start
+// to finish, including possible restart of already started process. In the latter case the
+// method should be just called again and it will pick up already started process. The only
+// input argument is the SQL statements that comprise the schema change that needs to be
+// pushed using the schema swap process.
+func (schemaSwap *Swap) Run(sql string) error {
+	err := schemaSwap.createShardObjects()
```
Comment (alainjobart) [inline go/vt/schemamanager/schemaswap/schema_swap.go]: I know we already mentioned this, this is more common in our code:
if err := ....; err != nil {
    return err
}

File: go/vt/schemamanager/schemaswap/schema_swap.go [hunk 2]
```diff
@@ -75,6 +90,290 @@ type shardSchemaSwap struct {
...
+	if err != nil {
+		return err
+	}
+	err = schemaSwap.initializeSwap(sql)
+	if err != nil {
+		return err
+	}
+	err = schemaSwap.runOnAllShards(
```
Comment (alainjobart) [inline go/vt/schemamanager/schemaswap/schema_swap.go]: How about using a different variable name than err, here? Like:
errHealth := ....
Then the if err != nil below would be a lot more readable.

File: go/vt/schemamanager/schemaswap/schema_swap.go [hunk 3]
```diff
@@ -75,6 +90,290 @@ type shardSchemaSwap struct {
...
+// the the method just picks up that. Otherwise it starts a new one and writes into the database that
+// the process was started.
+func (schemaSwap *Swap) initializeSwap(sql string) error {
+	metadataChannel := make(chan *shardSwapMetadata)
+	for _, shard := range schemaSwap.allShards {
+		go shard.readShardMetadata(metadataChannel)
+	}
+	var initError error
+	var cntDataReceived int
+	for cntDataReceived != len(schemaSwap.allShards) {
```
Comment (pivanof) [inline go/vt/schemamanager/schemaswap/schema_swap.go]: Done.

File: go/vt/schemamanager/schemaswap/schema_swap.go [hunk 4]
```diff
@@ -75,6 +90,290 @@ type shardSchemaSwap struct {
...
+	}
+	tabletInfo, err := topoServer.GetTablet(shardSwap.parent.ctx, shardInfo.MasterAlias)
+	if err != nil {
+		return nil, err
+	}
+	return tabletInfo.Tablet, nil
+}
+
+// readShardMetadata reads info about schema swaps on this shard from _vt.shard_metadata table.
+func (shardSwap *shardSchemaSwap) readShardMetadata(metadataChannel chan *shardSwapMetadata) {
```
Comment (pivanof) [inline go/vt/schemamanager/schemaswap/schema_swap.go]: I believe returning error here will hurt code compactness and readability. So I left it as is.

File: go/vt/schemamanager/schemaswap/schema_swap.go [hunk 5]
```diff
@@ -106,26 +106,24 @@ type shardSwapMetadata struct {
...
 	// succeed while others fail. We should try to stop health watching on all shards no
 	// matter if there was some failure or not.
 	defer schemaSwap.stopAllHealthWatchers()
-	if err != nil {
-		return err
+	if errHealthWatchers != nil {
+		return errHealthWatchers
 	}
-	err = schemaSwap.runOnAllShards(
+	err := schemaSwap.runOnAllShards(
```
Comment (alainjobart) [inline go/vt/schemamanager/schemaswap/schema_swap.go]: Let's also use:
if err := ...; err != nil {
below?

File: go/vt/schemamanager/schemaswap/schema_swap.go [hunk 6]
```diff
@@ -211,48 +202,52 @@ func (schemaSwap *Swap) stopAllHealthWatchers() {
...
-	var cntDataReceived int
-	for cntDataReceived != len(schemaSwap.allShards) {
-		metadata := <-metadataChannel
-		cntDataReceived++
-		if metadata.lastStartedSwap == metadata.lastFinishedSwap {
+	var waitGroup sync.WaitGroup
+	metadataList := make([]shardSwapMetadata, len(schemaSwap.allShards))
+	for i, shard := range schemaSwap.allShards {
+		waitGroup.Add(1)
+		go shard.readShardMetadata(&metadataList[i], &waitGroup)
```
Comment (alainjobart) [inline go/vt/schemamanager/schemaswap/schema_swap.go]: Alternate here, that may be a bit more readable, but up to you:
go func() {
  defer waitGroup.Done()
  shard.readShardMetadata(&metadataList[i])
}()

Comment (pivanof) [inline go/vt/schemamanager/schemaswap/schema_swap.go]: Yes, I agree it's more readable, and I've tried exactly this already. :) I've got these errors on commit:

go/vt/schemamanager/schemaswap/schema_swap.go:211: range variable shard captured by func literal
go/vt/schemamanager/schemaswap/schema_swap.go:211: range variable i captured by func literal

Please fix the go vet warnings above. To disable certain checks, change vetflags in misc/git/hooks/govet.

File: go/vt/schemamanager/schemaswap/schema_swap.go (hunk 5 above)
Comment (pivanof) [inline go/vt/schemamanager/schemaswap/schema_swap.go]: That will make "if" condition spread over 4 lines and blend with the code inside "if" due to the same indentation. Are you sure that needs to be done?

File: go/vt/schemamanager/schemaswap/schema_swap.go (hunk 6 above)
Comment (alainjobart) [inline go/vt/schemamanager/schemaswap/schema_swap.go]: Ah yes you'd have to pass the changing variables to the function, not worth it. Never mind.

File: go/vt/schemamanager/schemaswap/schema_swap.go [hunk 7]
```diff
@@ -75,6 +91,283 @@ type shardSchemaSwap struct {
...
+	if errHealthWatchers != nil {
+		return errHealthWatchers
+	}
+	err := schemaSwap.runOnAllShards(
+		func(shard *shardSchemaSwap) error {
+			return shard.applySeedSchemaChange(sql)
+		})
+	if err != nil {
+		return err
+	}
```
Comment (alainjobart) [inline go/vt/schemamanager/schemaswap/schema_swap.go]: I guess I would write this as:
if err := schemaSwap.runOnAllShards(func(shard *shardSchemaSwap) error {
    return shard.applySeedSchemaChange(sql)
}); err != nil {
    return err
}

but yeah, indentation makes it a bit weirder, so up to you.

Comment (pivanof) [inline go/vt/schemamanager/schemaswap/schema_swap.go]: Your suggestion is getting converted by gofmt into this:
<pre>
	if err := schemaSwap.runOnAllShards(func(shard *shardSchemaSwap) error {
		return shard.applySeedSchemaChange(sql)
	}); err != nil {
		return err
	}
</pre>
This is way beyond readable. I'll keep the code as is.

**Comment by alainjobart** [comment]
LGTM and agreed, the last proposed change is just not readable.

**Pull request merged**